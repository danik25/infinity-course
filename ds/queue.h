/****************************************************************
*                Data Structures - Queue                        *
* Related files: queue.c, queue_test.c                          *
* #Version: V 1.2                                               *
* Writer: Dani Benjamin                                         *
* Reviewer: Lev Shubert                                         *
* Last update: 10/04/19                                         *
****************************************************************/  

#ifndef _QUEUE_ /* INCLUDE_GUARD */
#define _QUEUE_
#include <stddef.h> /* size_t */

/*
 * Queue is an abstract data type or a linear data structure, just like stack
 * data structure, in which the first element is inserted from one end called  
 * the REAR(also called tail), and the removal of existing element takes place 
 * from the other end called as FRONT(also called head).This makes queue as  
 * FIFO(First in First Out) data structure, which means that element inserted  
 * first will be removed first.The difference between stacks and queues is in 
 * removing.In a stack we remove the item the most recently added; in a queue, 
 * difference between stacks and queues is in removing. In a stack we remove 
 * we remove the the item the most recently added; in a queue, we remove the  
 * item the least recently added. 
 */ 

/* API Design:
 *	
 *		This API design is an infrastructure module, allowing the user
 *		to access the structure.
 *		With the ability to access the structure, the user can add nodes,
 *		delete nodes, modify the structure itself as well as the data in
 *		any of it's fields.
 *
 *		This design supports any data type. 
 *
 */

typedef struct Queue queue_t;

/* 
 * Create a queue.
 * Return Value:  
 * if secceded -pointer to the created queue.
 * if failed - NULL. 
 */
queue_t *QCreate(void);

/* Destroy the queue- free all allocated memory.*/
void QDestroy(queue_t *queue);	  


/* 
 * Insert a member to the end of the queue.
 *  
 * Params: 
 *		  queue - the queue. 
 *		  data  - the data to be inserted.
 * Return value : 
 *        0-(success), 1-(failure).
 */
int QEnqueue(queue_t *queue, void *data);


/* 
 * Remove a member from the head of the queue.
 * returns a pointer to data
 */ 
void *QDequeue(queue_t *queue);


/* 
 * return a pointer to data at the head of the queue.
 * Return value : 
 *		  a pointer to the head's data. 
 */
void *QPeek(const queue_t *queue);


/* 
 * count number of elements in the queue
 *
 * Return value : 
 *       queue's length.
 */
size_t QSize(const queue_t *queue);


/* 
 * a boolean function that checks if queue has no elements.
 *	
 * Return value : 
 *		  1 - empty, 0 - not empty.
 */
int QIsEmpty(const queue_t *queue);


/* 
 * joins two queues together. free memory of <from> queue.
 * Params: 
 *		  to   - Will be at the head of the new queue.
 *		  from - will come after the <to> and will end the new queue.
 * Return value : 
 *		  a pointer to new appended queue.
 */
queue_t *QAppend(queue_t *to, queue_t *from);

#endif /* _QUEUE_ */
