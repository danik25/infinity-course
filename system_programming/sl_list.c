/****************************************************************
*                Data Structures - Singly Linked List           *
* Related files:                                                *
* #Version: V 1                                                 *
* writer: Dani Benjamin                                         *
* reviewer: Oded Bryt                                           *
* last update: 10.4.19                                          *
****************************************************************/ 

#include <stdlib.h> /* malloc */
#include <assert.h> /* assert */
#include "sl_list.h"


/****************************************************************/

sl_node_t *SLLCreate(void *data, sl_node_t *next)
{
    sl_node_t *sl_node = NULL;
    sl_node = (sl_node_t*) malloc(sizeof(sl_node_t));
    if (NULL == sl_node)
    {
        return NULL;
    }
    
    sl_node->data = data;
    sl_node->next = next;
    
    return (sl_node); 
}

/****************************************************************/

void SLLDestroy(sl_node_t *node)
{
    sl_node_t *tmp = NULL;
    
    while(node != NULL)
    {
        tmp = node;
        node = node->next;
        free(tmp);
    }
}

/****************************************************************/

sl_node_t *SLLInsert(sl_node_t *node_to_insert, sl_node_t *node_location)
{
	void *tmp = NULL;
	
	assert(node_location);
	
	/* saving the location data */
	tmp = node_location->data;
	
	/* switching the data between the inserted node and the location node */
	node_location->data = node_to_insert->data;
	node_to_insert->data = tmp;
	
	node_to_insert->next = node_location->next;
	node_location->next = node_to_insert;

	return (node_location);
}

/****************************************************************/

sl_node_t *SLLInsertAfter(sl_node_t *node_to_insert, sl_node_t *node_location)
{
    assert(node_to_insert);
    assert(node_location);
    
    node_to_insert->next = node_location->next;
    node_location->next = node_to_insert;
    
    return (node_to_insert); 
}

/****************************************************************/

sl_node_t *SLLRemove(sl_node_t *node_to_remove)
{
	sl_node_t *tmp_next = NULL;
	void *tmp = NULL;
	
	assert(node_to_remove);
	
	if(NULL == node_to_remove->next)
	{
		return (NULL);
	}
	
	/* switching the data between the removed node and the next one */
	tmp = node_to_remove->next->data;
	node_to_remove->next->data = node_to_remove->data;
	node_to_remove->data = tmp;
	
	/* detaching the pointer to the removed node */
	tmp_next = node_to_remove->next;
	node_to_remove->next = node_to_remove->next->next;
	tmp_next->next = NULL;
	
	return (tmp_next);
}

/****************************************************************/

sl_node_t *SLLRemoveAfter(sl_node_t *location)
{
	sl_node_t *tmp = NULL;
	
	assert(location);
	
	if(NULL == location->next)
	{
		return NULL;
	}

    tmp = location->next;
    location->next = location->next->next; 

    tmp->next = NULL;
    
    return (tmp);  
}
/****************************************************************/

int SLLForEach(sl_node_t *node, action_function_t func, void *param)
{
    int status = 0;
    
    assert(node);
    assert(func);
    
    while(node != NULL)
    {
        status = func(node->data, param);
        if(status)
        {
            return (status); /*fail*/
        }
        node = node->next;
    }
    return (status);
}

/****************************************************************/

sl_node_t *SLLFind(sl_node_t *node, is_match_function_t func, void *param)
{
    assert(func);
    while(node != NULL)
    {
        if(func(node->data, param))
        {
            return (node);
        }
        node = node->next;
    }
    return (NULL);
}

/****************************************************************/

size_t SLLCount(const sl_node_t *node)
{
	size_t cnt = 0;
	
	if(NULL == node)
	{
		return (0);
	}
	
	while(node != NULL)
	{
		node = node->next;
		++cnt;
	}
	
	return (cnt);
}

/****************************************************************/
sl_node_t *SLLFlip(sl_node_t *node)
{
    sl_node_t *tmp = NULL;
	sl_node_t *tmp_next = NULL;
	
	assert(node);
	
	while(node)
	{
	    /* appending pointers */
	    tmp = tmp_next;
	    tmp_next = node;
	    node = node->next;
	    
	    /* assign "next" to "previous" */
	    tmp_next->next = tmp;
	}
	
	node = tmp_next;
	
	return (node);
}

/****************************************************************/

sl_node_t *SLLFindIntersection(const sl_node_t *node_1, 
       						   const sl_node_t *node_2)
{
	int gap = 0;
	
	assert(node_1);
	assert(node_2);
	
	gap = SLLCount(node_1) - SLLCount(node_2);
	if(gap > 0)
	{
		while(gap > 0)
		{
			node_1 = node_1->next;
			--gap;
		}
	}
	else
	{
		gap *= -1;
		while(gap > 0)
		{
			node_2 = node_2->next;
			--gap;
		}
	}
	while(node_1 != NULL)
	{
		if(node_1 == node_2)
		{
			return ((sl_node_t *)node_1);
		}
		
		node_1 = node_1->next;
		node_2 = node_2->next;
	}
	return (NULL);	
}

/****************************************************************/

int SLLHasLoop(sl_node_t *node)
{
	sl_node_t *by_one = NULL;
	sl_node_t *by_two = NULL;
	
	assert(node);
	
	by_one = node;
	by_two = node->next;
	
	while (by_one && by_two && by_two->next)
	{
		if(by_one == by_two)
		{
			return (1);
		}
		by_one = by_one->next;
		by_two = by_two->next->next;
	}
	return (0);
}


