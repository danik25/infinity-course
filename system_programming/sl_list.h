/****************************************************************
*                Data Structures - Singly Linked List           *
* Related files:                                                *
* #Version: V 1                                                 *
* writer: Dani Benjamin                                         *
* reviewer: Oded Bryt                                           *
* last update: 10.4.19                                          *
****************************************************************/

#ifndef _SL_LIST_
#define _SL_LIST_
#include <stddef.h> /* size_t */
 
/*
 * A singly linked list is a data structure consisting of a group of nodes,
 * linked together. Each node consists of data elements and a link
 * to the next node. The main advantage of linked lists over simple arrays 
 * is that the nodes can be inserted anywhere in the list without reallocating
 * the entire array because the nodes in a linked list do not need to be stored
 * contiguously in memory. However, nodes in the list cannot be accessed by 
 * index but must be accessed by traversing the list.
*/  

/* API Design:
	
		This API design is an infrastructure module, allowing the user
		to access the structure.
		With the ability to access the structure, the user can add nodes,
		delete nodes, modify the structure itself as well as the data in
		any of it's fields.

		This design supports any data type. 
	

	******************     *****************    *****************    
  	* DATA  *   next * --> *  DATA  * next * -> *  DATA  * next *---> NULL 
	******************     *****************    *****************
	   	  1st node 		    2nd node  ....n'th node...    Last node	
*/


/* typedef on function pointer for readability - both functions are given */
/* as arguments to SLLForEach and SLLFind respectively. */
typedef int (action_function_t)(void *data, void* param);
typedef int (is_match_function_t)(void *data, void* param);

 
typedef struct sl_node {
		    	  	
		void * data;
		struct sl_node *next; 

} sl_node_t;



/* Create a node, can connect the new node to other existing nodes.

 * Params:
		  next - pointer to the node that will be pointed by the created node.
				 (NULL if node does not point to any other node).
		  data - pointer to the data that will be inserted to the node.
 
 * Return Value: pointer to the created node. 
*/

sl_node_t *SLLCreate(void *data, sl_node_t *next);



/* Destroy the linked list - free all allocated memory from a given position.

 * Params: 
		  node - reference node.

 * Time complexity: O(n).
*/
void SLLDestroy(sl_node_t *node);	  



/* Insert a node before a given location.

 * Params: 
		  node_to_insert - desired node to be inserted. 
		  node_location  - node that will be pointed to by the inserted node.
		  

 * Return value : pointer to inserted node
*/

sl_node_t *SLLInsert(sl_node_t *node_to_insert, sl_node_t *node_location);



/* Insert a node after a given location.

 * Params: 
		  node_location  - the node that will point to the inserted node.
		  node_to_insert - the node to be inserted.

 * Return value : pointer to inserted node. 
*/

sl_node_t *SLLInsertAfter(sl_node_t *node_to_insert, sl_node_t *node_location);



/* Remove a node from a given location.Cannot remove the last node from a list.

 * Params: 
		  node_to_remove - the node that will be removed from list.

 * Return value : a pointer to the removed node.
*/
sl_node_t *SLLRemove(sl_node_t *node_to_remove);


/* Remove a node from the list after a given location.

 * Params: 
		  location  - the node that points to the node to be removed.

 * Return value : a pointer to the removed node. 
*/
sl_node_t *SLLRemoveAfter(sl_node_t *location);


/* Apply <action_func> on all nodes starting from <node>.

 * Params: 
		  node - reference node.

 * Return value : success / type of failure.
*/
int SLLForEach(sl_node_t *node, action_function_t func, void *param);



/* Find if a certain data matches data on the list.

 * Params: 
		  node - reference node.
		  data - the data to be found.	

 * Return value : pointer to node that holds relevant data.
*/
sl_node_t *SLLFind(sl_node_t *node, is_match_function_t func, void *param);


/* Count the number of nodes on the list, from a reference node.

 * Params: 
		  node - reference node.

 * Return value : number of nodes on the list.
*/
size_t SLLCount(const sl_node_t *node);



/* Reverse the direction of the singly linked list.

 * Params: 
		  node - reference node.

 * Return value : a pointer to the new first node of the list.
*/
sl_node_t *SLLFlip(sl_node_t *node);


/* Find intersections - a shared node between two lists.

 * Params: 
		  node_1 - reference node of first list.
  		  node_2 - reference node of second list.

 * Time complexity: O(n).

 * Return value : a pointer to the shared node.
*/
sl_node_t *SLLFindIntersection(const sl_node_t *node_1, 
       						   const sl_node_t *node_2);



/* Finds if a loop exists within the list.

 * Params: 
		  node  - reference node.
	
 * Return value : 1 if loop found, 0 if not.
*/
int SLLHasLoop(sl_node_t *node);


#endif /* _SL_LIST_ */
