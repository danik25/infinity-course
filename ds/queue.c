/****************************************************************
*                Data Structures - Queue                        *
* Related files: queue.c, queue_test.c                          *
* #Version: V 1.2                                               *
* Writer: Dani Benjamin                                         *
* Reviewer: Lev Shubert                                         *
* Last update: 10/04/19                                         *
****************************************************************/

#include <stdlib.h> /* malloc */
#include <assert.h> /* assert */
#include "queue.h"
#include "sl_list.h"


struct Queue {	    	  	
				sl_node_t *head;	
				sl_node_t *tail;

};

queue_t *QCreate(void)
{
    queue_t *new_q = NULL;
    size_t address = 0;
    new_q = (queue_t*) malloc(sizeof(queue_t));
    address = 0xDEADBEEF;
    if(!new_q)
    {
        return (NULL);
    }
    new_q->head = SLLCreate((void*)address, NULL);
    new_q->tail = new_q->head;
    
    return (new_q);    
}

void QDestroy(queue_t *queue)
{
    assert(queue);
    SLLDestroy(queue->head);
    free(queue);
    queue = NULL;
}

int QEnqueue(queue_t *queue, void *data)
{
    sl_node_t *tmp = NULL;
    
    assert(queue);
    
    tmp = SLLCreate(data, NULL);
    if(!queue->tail)
    {
        return (1); /* fail */ 
    }
  
    queue->tail = SLLInsert(tmp, queue->tail);
	queue->tail = queue->tail->next;
	
    return (0); /* success */
}

void *QDequeue(queue_t *queue)
{
	sl_node_t *answer = NULL;
    void *answer_data = NULL;
    
    assert(queue);
    
	if(queue->head == queue->tail)
	{
		return (NULL);
	}

    if(1 == QSize(queue))
    {
    	queue->tail = queue->head;
    }
    answer_data = queue->head->data;
    answer = SLLRemove(queue->head);
    
    SLLDestroy(answer);
    
    answer = NULL;
    
    return (answer_data);
}

void *QPeek(const queue_t *queue)
{
	void *data = NULL;
	
	assert(queue);
	
	data = queue->head->data;
	
	if(data != (void*)0xdeadbeef)
	{
		return (data);
	}
	return (NULL);
}

int QIsEmpty(const queue_t *queue)
{
	assert(queue);
	return (queue->head == queue->tail);
}

size_t QSize(const queue_t *queue)
{
	assert(queue);
	return (SLLCount(queue->head) - 1);
}

queue_t *QAppend(queue_t *to, queue_t *from)
{
	assert(to);
	assert(from);
	
	if(QIsEmpty(from))
	{
	    QDestroy(from);
	    return (to);
	}

	to->tail->next = from->head;
	
	free(SLLRemove(to->tail));
	to->tail = from->tail;
	
	free(from);
	
	return (to);
}







