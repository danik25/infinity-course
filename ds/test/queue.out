/home/xyz/git/dani-benjamin/ds/queue.out